### DISCUSSION TITLE
Custom action on code fences but keep the normal behavior

### DISCUSSION THREAD

**Discussion by oprypin** [facelessuser/pymdown-extensions#1291] [Q&A] [answered]
Hi. I want to validate code blocks in my MkDocs site (e.g. by giving them to an external process). The validation would basically have the ability to abort building the site. A further ability that I'd like to have is to modify the code snippet before passing it onwards.

Other than that, there should be no changes to what code fences do, and **normal syntax highlighting should continue to work**.
Ideally I'd implement a reusable plugin, not a one-off.

---

So I started looking, how *do* I get the content (and class) of code fences?

* I don't want to reimplement the parsing already done by SuperFences.
* A treeprocessor (after *superfences* processing) can't work, because it just sees the `htmlStash` placeholders.
* Looking at `htmlStash` itself is weird because it's already HTML. Also it has already discarded classes. It could kinda work, though.
* Should I monkeypatch/wrap the `highlight` function (through `treeprocessors["fenced_code_block"]`)? Really nasty but happens to satisfy all my requirements.
   https://github.com/facelessuser/pymdown-extensions/blob/390b15d6e23596b4606156741eebb190b91edae8/pymdownx/superfences.py#L739

---

Oh wait, ["custom fences"](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/#custom-fences) is the right feature!
* Should I do my check inside the `validator` and just always return `False`? It would probably work OK but
   * I have no ability to modify the code snippet
   * Requires specifying the "custom fences" config manually (but ideally I want a fully code-based extension).
     * Or is it feasible to somehow reach this function from my code? https://github.com/facelessuser/pymdown-extensions/blob/390b15d6e23596b4606156741eebb190b91edae8/pymdownx/superfences.py#L242 Would I need to look through `registeredExtensions` or something?
* Should I monkeypatch/wrap the specific item `SuperFencesCodeExtension.superfences[0]`?
  https://github.com/facelessuser/pymdown-extensions/blob/390b15d6e23596b4606156741eebb190b91edae8/pymdownx/superfences.py#L316 
* Should I extract the existing `SuperFencesBlockPreprocessor` just to get to its `highlight` method and use it as the formatter?
* Should I instantiate another `SuperFencesBlockPreprocessor` with manual config and use its `highlight` method?

---

Other options?

Basically my big issue with the supported approach is that there's no good way to still use the default `highlight` behavior if I'm overriding it with my "custom fence" action.

Or, well, going back to the overall problem, I basically just want a hook to be called on every fenced code block, without abandoning the highlighting behavior.

**Comment by facelessuser** [comment, marked as answer]
Yeah, you can use custom fences and call into the default formatter. I basically do that in this project: https://github.com/facelessuser/coloraide/blob/master/tools/colors.py#L210.

Basically, the above code is allowing me to parse raw Python code and get it's results, then I format it as a `pycon` input/output, send it through the highlighter and also about color previews:

````py
```color
Color("rebeccapurple").interpolate(
    "lch(85% 100 85)",
    space='lch'
)
```
````

<img width="746" alt="Screen Shot 2021-03-28 at 3 23 46 PM" src="https://user-images.githubusercontent.com/1055125/112768527-a4c54900-8fd9-11eb-9f5e-6a2968b934f3.png">

Generally, it is assumed that if someone is overriding this, they don't want to do pygments parsing, or they are going to implement their own; otherwise, they would normally just use the default highlighter. But yeah, I ran into a situation where I wanted to do much more, but still, leverage the highlighter. Custom fence usage is one of those things that has kind of evolved over time. It's one of those things, that given time, I push a bit further.

**Reply by oprypin** [reply]
Thanks!
So then, the expression `md.preprocessors['fenced_code_block'].extension.superfences[0]['formatter']` has some level of backwards compatibility guarantee :smile:

**Reply by facelessuser** [reply]
Yeah, some at least 🙂.

**Reply by oprypin** [reply]
Here's what I did with this:

[MkDocs plugin](https://github.com/oprypin/mkdocs-code-validator)
* [Injecting config into pymdownx.superfences](https://github.com/oprypin/mkdocs-code-validator/blob/600f82515835e2662c731070f91bf6664f04258b/mkdocs_code_validator/plugin.py#L83)
* [Fallback to [0]['formatter']](https://github.com/oprypin/mkdocs-code-validator/blob/600f82515835e2662c731070f91bf6664f04258b/mkdocs_code_validator/plugin.py#L135)

**Reply by facelessuser** [reply]
Very cool. Glad you were able to build something useful with it.
